posts/2022/05/the-evolution-of-the-scipy-developer-cli.md

```diff
@@ -0,0 +1,170 @@
+<!--
+.. title: The evolution of the SciPy developer CLI
+.. slug: the-evolution-of-the-scipy-developer-cli
+.. date: 2022-05-03 8:00:00 UTC+05:30
+.. author: Sayantika Banik
+.. tags: CLI, SciPy, doit, Typer, Rich, Click
+.. category:
+.. link:
+.. description:
+.. type: text
+.. previewimage:
+-->
+
+#### 🤔 What is a command-line interface (CLI)?
+
+Imagine a situation where there is a massive system with various tools and functionality. Every functionality requires a special command or an input from the user, CLI is designed to tackle such situations. Like a catalog it lists all the options present, helping the user to navigate.
+
+<p align="center">
+  <img alt="CLI example" src="https://user-images.githubusercontent.com/17350312/166633508-a2795c44-30bc-4a5b-8043-65beab71d31f.png" />
+</p>
+
+Now that we understand what a `CLI` is, how about we dive into the world of `SciPy`?
+
+<!-- TEASER_END -->
+
+#### 🤓 SciPy CLI journey
+
+An open-source project undergoes multiple iterations, and each iteration adds layers of tools and functionalities to incorporate. Right from the initial build, test, benchmarking, release notes, and beyond. Contribution guidelines start spanning across multiple pages, exponentially increasing efforts to maintain.
+
+The two homegrown CLI for SciPy are `runtests.py`(for distutils-based builds) and `dev.py`(for Meson-based builds) built using python tooling `argparse`.
+Both of these toolings have multiple conditional blocks to achieve the desired outcome. Not only it is difficult to maintain, overall readability decreases drastically. The documentation runs into an infinite loop of updates, requiring additional efforts from maintainers. Another issue that remains is the lack of task grouping. A group of tasks linking to a single objective helps reduce confusion, super helpful for new contributors (like me :D).
```

Comment:
Is `toolings` SciPy lingo or some term I'm not aware of? if not I word amend it to `tools`. 
the context for mentioning the conditionals isn't super clear to me. Why is this piece of  information interesting/relevant in an already information heavy paragraph?  If you think it's interesting for people who have knowledge I lack, do keep it. If you think it's interesting for new contributors , maybe add some insight as to why, like you do so well later on. If it's not critical, suggest to edit to `Both of these tools are hard to maintain, their overall readability decreases  ...`

Reply:
It's not a SciPy lingo (maybe just me using the term), I shall update the wording. Thanks for pointing :)

By conditional blocks, I was trying to infer multiple `if-else` statements. Maybe replacing the term `conditional` with the `if-else` along with an example could be helpful?

Reply:
Oh yeah, I understand that the conditionals mean if-end statements. I just don't understand why that information is relevant to the story you are telling me. Is it particularly interesting? It just seems like a very specific detail to mention here that isn't connected to the rest of the paragraph. Are conditionals notoriously hard to maintain and I'm missing the point?

Reply:
Long chains of conditional statements degrade code readability. Makes it harder to find the right block to update/modify something. 

Each conditional block is designed to pick a specific user input and perform certain actions. 
`if user_input==--test ...perform x`
     `else:___`
 `if user_input==--bench ...perform y`
      `else:___`
  This spans to multiple other blocks.
  
  I can remove the conditional wording or try adding a similar example.

Reply:
Ah, the way you explain it now is much more clear to me than the original text. May I suggest you amend to incorporate your new phrasing:
`Both of these tools have long chains of conditional statements, **which** are notorious for reducing code readability drastically. **That** makes it difficult to maintain such code, because it is harder to find the right block to modify.`

Now that you explained I can also understand the original 2 lines easily tbh. But originally, I didn't get the connection between  the first line and the second line.  In the suggestion I made I try to make the connection a bit stronger. so the reader understands the conditionals are the cause of the problem. But feel free to rewrite in your style.